internal/configuration/configuration.go

```diff
@@ -0,0 +1,313 @@
+	OidcIssuerEndpoint   string   `mapstructure:"oidc_issuer_endpoint"`
+	OidcAudienceEndpoint string   `mapstructure:"oidc_audience_endpoint"`
+	OidcClientSecret     string   `mapstructure:"oidc_client_secret"`
+	OidcClientID         string   `mapstructure:"oidc_client_id"`
+	OidcClientScopes     []string `mapstructure:"oidc_client_scopes"`
+	DisableOAuth         bool     `mapstructure:"disable_oauth"`
+}
+
+// Load the application configuration
+//
+// Reads in the cfgFile when available and overrides from environment variables.
+func Load(args *model.Args) (*Configuration, error) {
+	viperConfig := viper.New()
+	viperConfig.SetConfigType("yaml")
+	viperConfig.SetEnvPrefix(model.AppName)
+	viperConfig.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+	viperConfig.AutomaticEnv()
+
+	if args.ConfigFile != "" {
+		fh, err := os.Open(args.ConfigFile)
+		if err != nil {
+			return nil, errors.Wrap(model.ErrConfig, err.Error())
+		}
+
+		if err = viperConfig.ReadConfig(fh); err != nil {
+			return nil, errors.Wrap(model.ErrConfig, "ReadConfig error: "+err.Error())
+		}
+	}
+
+	config := New()
+	config.LoadArgs(args)
+
+	if err := config.envBindVars(viperConfig); err != nil {
+		return nil, errors.Wrap(model.ErrConfig, "env var bind error: "+err.Error())
+	}
+
+	if err := viperConfig.Unmarshal(config); err != nil {
+		return nil, errors.Wrap(model.ErrConfig, "Unmarshal error: "+err.Error())
+	}
+
+	config.envVarAppOverrides(viperConfig)
+
+	if err := config.envVarNatsOverrides(viperConfig); err != nil {
+		return nil, errors.Wrap(model.ErrConfig, "nats env overrides error: "+err.Error())
+	}
+
+	if err := config.envVarFleetDBOverrides(viperConfig); err != nil {
+		return nil, errors.Wrap(model.ErrConfig, "fleetdb env overrides error: "+err.Error())
+	}
+
+	return config, nil
+}
+
+func (c *Configuration) envVarAppOverrides(viperConfig *viper.Viper) {
+	logLevel := viperConfig.GetString("log.level")
+	if logLevel != "" {
+		c.LogLevel = logLevel
+	}
+}
+
+// envBindVars binds environment variables to the struct
+// without a configuration file being unmarshalled,
+// this is a workaround for a viper bug,
+//
+// This can be replaced by the solution in https://github.com/spf13/viper/pull/1429
+// once that PR is merged.
+func (c *Configuration) envBindVars(viperConfig *viper.Viper) error {
+	envKeysMap := map[string]interface{}{}
+	if err := mapstructure.Decode(c, &envKeysMap); err != nil {
+		return err
+	}
+
+	// Flatten nested conf map
+	flat, err := flatten.Flatten(envKeysMap, "", flatten.DotStyle)
+	if err != nil {
+		return errors.Wrap(err, "Unable to flatten configuration")
+	}
+
+	for k := range flat {
+		if err := viperConfig.BindEnv(k); err != nil {
+			return errors.Wrap(model.ErrConfig, "env var bind error: "+err.Error())
+		}
+	}
+
+	return nil
+}
+
+// nolint:gocyclo // nats env configuration load is cyclomatic
+func (c *Configuration) envVarNatsOverrides(viperConfig *viper.Viper) error {
```

Comment:
A lot of these NATS configuration parameters don't have to be loaded with the rivets.Controller lib anymore - https://github.com/metal-toolbox/flasher/blob/56b74a06018ba46b38bd82ccd48eaf531f7571fc/internal/app/config.go#L159

Reply:
Good! Initially I didn't mess with the NATS config because I wanted to it just work. Done.